.github/workflows/main.yml

```diff
@@ -0,0 +1,65 @@
+name: Tests and static analysis
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - main
+
+env:
+  CARGO_TERM_COLOR: always
+
+jobs:
+  check:
+    name: Check
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest]
+        kind: [debug, release]
+    runs-on: ${{ matrix.os }}
+    env:
+      RUSTFLAGS: "-D warnings"
+    steps:
+      - uses: actions/checkout@v4
+      - run: rustup update
+      - run: cargo check ${{ matrix.kind == 'release' && '--release' || '' }}
+      - run: cargo clippy ${{ matrix.kind == 'release' && '--release' || '' }}
+
+  test:
+    name: Run tests
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest]
+        kind: [debug, release]
+    runs-on: ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v4
+      - run: rustup update
```

Comment:
Is there a reason why you would manually run `rustup update` instead of using the [dtolnay/rust-toolchain](https://github.com/dtolnay/rust-toolchain) action? The latter option would make some additional setup changes (like enabling sparse checkouts of the cargo registry, disabling incremental compilation and enabling color output (even though you already did that manually).

Reply:
Ah, I wasn't aware of the additional benefits of that action, thanks for pointing it out. I'll switch to using it, then.